pkg/blobinfocache/internal/prioritize/prioritize.go

```diff
@@ -77,34 +82,58 @@ func (css *candidateSortState) Swap(i, j int) {
-// number of entries to limit, only to make testing simpler.
-func destructivelyPrioritizeReplacementCandidatesWithMax(cs []CandidateWithTime, primaryDigest, uncompressedDigest digest.Digest, maxCandidates int) []blobinfocache.BICReplacementCandidate2 {
+// destructivelyPrioritizeReplacementCandidatesWithMax is destructivelyPrioritizeReplacementCandidates with parameters for the
+// number of entries to limit for known and unknown location separately, only to make testing simpler.
+func destructivelyPrioritizeReplacementCandidatesWithMax(cs []CandidateWithTime, primaryDigest, uncompressedDigest digest.Digest, maxCandidates int, unknownLocationLimit int) []blobinfocache.BICReplacementCandidate2 {
+	// split unknown candidates and known candidates
+	// and limit them seperately.
+	var knownLocationCandidates []CandidateWithTime
+	var unknownLocationCandidates []CandidateWithTime
+
+	for _, candidate := range cs {
+		if candidate.Candidate.UnknownLocation {
+			unknownLocationCandidates = append(unknownLocationCandidates, candidate)
+		} else {
+			knownLocationCandidates = append(knownLocationCandidates, candidate)
+		}
+	}
+
 	// We don't need to use sort.Stable() because nanosecond timestamps are (presumably?) unique, so no two elements should
 	// compare equal.
 	sort.Sort(&candidateSortState{
-		cs:                 cs,
+		cs:                 knownLocationCandidates,
 		primaryDigest:      primaryDigest,
 		uncompressedDigest: uncompressedDigest,
 	})
 
-	resLength := len(cs)
+	resLength := len(knownLocationCandidates)
 	if resLength > maxCandidates {
 		resLength = maxCandidates
 	}
 	res := make([]blobinfocache.BICReplacementCandidate2, resLength)
 	for i := range res {
-		res[i] = cs[i].Candidate
+		res[i] = knownLocationCandidates[i].Candidate
+	}
+	// If candidates with unknown location are found, lets add them to final list
+	if len(unknownLocationCandidates) > 0 {
+		resUnknownLength := len(unknownLocationCandidates)
+		if resUnknownLength > unknownLocationLimit {
+			resUnknownLength = unknownLocationLimit
+		}
+		for i := 0; i < resUnknownLength; i++ {
+			res = append(res, unknownLocationCandidates[i].Candidate)
+		}
 	}
 	return res
 }
 
 // DestructivelyPrioritizeReplacementCandidates consumes AND DESTROYS an array of possible replacement candidates with their last known existence times,
-// the primary digest the user actually asked for, and the corresponding uncompressed digest (if known, possibly equal to the primary digest),
-// and returns an appropriately prioritized and/or trimmed result suitable for a return value from types.BlobInfoCache.CandidateLocations.
+// the primary digest the user actually asked for, the corresponding uncompressed digest (if known, possibly equal to the primary digest) and Location
+// (where unknown location is always given lower priority as compared to candidates with known locations) and returns an appropriately prioritized and/or
```

Comment:
```suggestion
// (which might be unknown) and returns an appropriately prioritized and/or
```
_how_ this function prioritizes is not an API promise.